src/arr/compiler/anf-loop-compiler.arr

```diff
@@ -520,6 +520,37 @@ var total-time = 0
 
 show-stack-trace = false
 fun compile-fun-body(l :: Loc, step :: A.Name, fun-name :: A.Name, compiler, args :: List<N.ABind>, opt-arity :: Option<Number>, body :: N.AExpr, should-report-error-frame :: Boolean, is-flat :: Boolean, is-method :: Boolean) -> J.JBlock block:
+  var in-lam = false
+  var arg-used-in-lambda = false
+  arg-names = args.map(_.id)
+  dummy-anf-lettable = N.a-obj(A.dummy-loc, empty)
+  body.visit(N.default-map-visitor.{
+    method a-lam(self, _, _, _, _, shadow body) block:
+      saved-in-lam = in-lam
+      in-lam := true
+      body.visit(self)
+      in-lam := saved-in-lam
+      dummy-anf-lettable
+    end,
+    method a-method(self, _, _, _, _, shadow body) block:
+      saved-in-lam = in-lam
+      in-lam := true
+      body.visit(self)
+      in-lam := saved-in-lam
+      dummy-anf-lettable
+    end,
+    method a-id(self, shadow l, id) block:
+      when in-lam and not(arg-used-in-lambda) and arg-names.member(id):
+        arg-used-in-lambda := true
+      end
+      N.a-id(l, id)
+    end
+  })
```

Comment:
I _think _you might need another case here, to handle `s-case-branch`es.  We sometimes compile those to be lambdas rather than inlined switch cases, and those might close over arguments too.  Definitely should test this situation to find out.